specs/interop.md

```diff
@@ -0,0 +1,188 @@
+# Interoperability
+
+The ability for a blockchain to easily read the state of another blockchain is called interoperability.
+Low latency interoperability allows for horizontally scalable blockchains, a key feature of the superchain.
+
+Note: this document references an "interop network upgrade" as a temporary name. The actual name of the
+network upgrade will be included in this document in the future.
+
+## Overview
+
+| Term          | Definition                                               |
+|---------------|----------------------------------------------------------|
+| Local Chain   | A blockchain that includes an initiating message |
+| Remote Chain  | A blockchain that includes a relaying message |
+| Initiating Message | A transaction submitted to a local chain that authorizes execution on a remote chain |
+| Relaying Message | A transaction submitted to a remote chain that corresponds to an initiating message |
```

Comment:
Do we allow relaying messages that relay generic "block header" information (state-root, maybe receipts root, etc.), rather than an actual initiated message? I think the fault-proof can support that easily, since outgoing initiating messages can be accumulated into the block-header, and that's all accessible in the current proposed consolidation step.
This would allow state-reads and such. The relaying message pays for relay anyway.

Reply:
This is a very interesting idea and am very curious about what can be built with it. Do you have any specific examples of things that can be built with it? Just want to make sure we are going in the straightest line towards so we need less testing and edge cases.

I could certainly imagine a type of withdrawal that isn't a call but instead "teleports" the block header into a remote chain. For what its worth, its possible to do this not natively at the application layer with this spec. You create a contract on both sides, the local contract uses the blockhash opcode and sends it across, the remote is `onlyOtherContract` so nobody else can store hashes in it and it stores the blockhash, then the user unwraps the blockhash with the header.

I think there is something interesting here with regards to credible commitments, ie sequencer of chain A includes a payment to bob and then a proof of that is sent to another chain where the other chain can statically analyze the tx (unwrapped from blockhash) but its still a bit researchy. Our sequencer MEV ecosystem isnt mature enough for this.

I propose that we don't bake this feature into the protocol directly now because it can be accomplished at the app layer and then if people do begin to use this functionality at the app layer then we can "precompile" or "nationalize" the functionality by pulling it into the protocol

Reply:
@tynes this was the original reason for collecting the `roots` in the the v2 implementation. Not interop critical code but unlocks immediate cross-chain state reads.

However this does introduce the system transaction it looks like v3 is trying to avoid

Reply:
hmm oh @protolambda, are you suggesting relaying generic header information external to the protocol? I wonder what party would be incentivized to this for free

Reply:
> but unlocks immediate cross-chain state reads

This can be built at the application layer, we should be careful about adding scope to the protocol unless we know for a fact it solves a problem

Reply:
> hmm oh @protolambda, are you suggesting relaying generic header information external to the protocol? I wonder what party would be incentivized to this for free

I interpreted this as internal to the protocol. I wouldn't ask the question "what party would be incentivized to do this for free", I would ask the question "what is a concrete usecase for this that is unlocked by baking it into the protocol instead of having it at the application layer?"